src/core/webview/ClineProvider.ts

```
@@ -947,11 +947,47 @@ export class ClineProvider extends EventEmitter<ClineProviderEvents> implements
 	}
 
 	async updateCustomInstructions(instructions?: string) {
-		// User may be clearing the field.
-		await this.updateGlobalState("customInstructions", instructions || undefined)
+		const settingsDirPath = await this.ensureSettingsDirectoryExists()
+		const customInstructionsFilePath = path.join(settingsDirPath, GlobalFileNames.customInstructions)
+
+		if (instructions && instructions.trim()) {
+			await fs.writeFile(customInstructionsFilePath, instructions.trim(), "utf-8")
+		} else {
+			// If instructions are empty or undefined, delete the file if it exists
+			try {
+				await fs.unlink(customInstructionsFilePath)
+			} catch (error) {
+				// Ignore if file doesn't exist
+				if ((error as NodeJS.ErrnoException).code !== "ENOENT") {
+					throw error
+				}
+			}
+		}
 		await this.postStateToWebview()
 	}
 
+	async openCustomInstructionsFile(): Promise<void> {
+		const settingsDirPath = await this.ensureSettingsDirectoryExists()
+		const customInstructionsFilePath = path.join(settingsDirPath, GlobalFileNames.customInstructions)
+		const fileUri = vscode.Uri.file(customInstructionsFilePath)
+		await vscode.commands.executeCommand("vscode.open", fileUri, { preview: false, preserveFocus: true })
+		await vscode.commands.executeCommand("workbench.action.files.revert", fileUri)
+	}
+
+	async refreshCustomInstructions(): Promise<void> {
+		const settingsDirPath = await this.ensureSettingsDirectoryExists()
+		const customInstructionsFilePath = path.join(settingsDirPath, GlobalFileNames.customInstructions)
+		let content: string | undefined = undefined
+		try {
+			content = await fs.readFile(customInstructionsFilePath, "utf-8")
+		} catch (error) {
+			if ((error as NodeJS.ErrnoException).code !== "ENOENT") {
+				throw error
```

I noticed we're silently catching ENOENT errors here but throwing others. Is this intentional? Should we be consistent with error handling across all the new methods?

---

4b803f034d2d83d618abdeadcfe0315f3a62ca74